### ISSUE TITLE
Don't commit the session in view functions

### ISSUE THREAD

**Issue by lukasjuhrich** [agdsn/pycroft#136]
There are several (>10) instances in `web.blueprints` where a manual commit of the session is issued. Such a manual intervention should not be necessary as this is the responsibility of the library `lib`.

**Comment by sebschrader** [comment]
I don't necessarily agree. The HTTP request context/lifetime should also resemble the lifetime of the SQL transaction.

Abstracting transactions away from the user of the library API is not the right thing to do. This is not an implementation detail, it should rather be treated first class. The library doesn't know how many different updates the caller wants to perform in one transaction, only the user of the library does. The library might start it's own nested transactions.

In my opinion every endpoint that performs changes, should contain a with block:

```python
with session.begin():
   …
```

This gets even more important, if we enable higher isolation levels, in which transaction might fail due to locking issues. In this case, we might want to perform automatic retries.

One could certainly move most of it out into decorator and/or context manager, so that the simple endpoints don't have to implement his boilerplate on their own every time.

**Comment by lukasjuhrich** [comment]
Let me phrase this differently.

There should currently be no **reason** for the view functions to wrap every library call with a `session.begin()`, because the library itself does not a priori know whether it is being called as a subroutine inside another library function (e.g. `log_user_event`), or resembles the top of the context stack. Most of the endpoints just resemble an interface to a particular library function such as `move_in`. IMO, it should not be necessary to set up a session just to make a lib call have an effect.
_however_, I totally agree with you that we probably should, independently of that, use a wrapping transaction anywhere to fit the request lifetime.

However, if something should actually break when removing such a `session.begin()`, this should be fixed.

**Comment by sebschrader** [comment]
I don't agree. To quote [The Zen of Python](https://www.python.org/dev/peps/pep-0020/):

> Explicit is better than implicit.

I would actually prefer, that users of the library have to manage the session themselves, instead of all this implicit business. In this case, the lib functions could take the session as an argument, instead of importing a magic global session object. This would clearly indicate, who is responsible for managing the session. The user of the library must already set up this magic global session somehow beforehand, so she is responsible anyway, it's just hidden and implicit.

The library can then stop wrapping everything in transactions, except where it really wants to use  savepoints, because it's not its job anymore to manage transactions (it never should have been in the first place).

I know, this is what Flask does for its `request`, `session`, `app` etc. objects, but it has been criticized for exactly that (and IMHO rightly so). We won't and shouldn't change how Flask works, but for our code, we can decide, how we want to do things.

We get rid of most of the boilerplate, by using a decorator:

```python
@wrapt.decorator
def foo(wrapped, instance, args, kwargs):
    session = <obtain session>
    with session.begin()
        return wrapped(session, *args, **kwargs)

@app.route('/bar/<identifier:int>')
@foo
def (session, identifier):
   baz = lib.do_bar_stuf(session, identifier)
   return render_template('bar.html', baz=baz)
```

**Comment by Georgeto** [comment]
> This gets even more important, if we enable higher isolation levels, in which transaction might fail due to locking issues. In this case, we might want to perform automatic retries.

Have we already assessed whether we have to enable high isolation levels? I highly doubt it as I was not able to find any code that explicitly sets a transaction isolation level. Although some of the more complex operations would definitely need a higher isolation level than PostgreSQL's default `READ COMMITED`.

**Comment by lukasjuhrich** [comment]
Over time I've come to agree with @sebschrader, and in particular [this section in the docs](https://docs.sqlalchemy.org/en/14/orm/session_basics.html#session-faq-whentocreate) may be enlightening for anyone reading.

- calling `session.begin` inside a lib function is bad because that only makes sense when the session has not begun yet, it should probably be `begin_nested` and only used when we actually need a savepoint
- calling `session.commit` inside a lib function without a corresponding `begin` is bad because it commits the outer session, and prohibits us from rolling it back if we want to (e.g. because the test which called it finished).

**Comment by lukasjuhrich** [comment]
What about the following steps:
- [ ] deprecate the `@with_transaction` decorator to favor explicit `session.begin` in the view function
- [ ] deprecate usage of the `session` proxy  to favor dependency injection, i.e. passing the session explicitly as `lib.foo(session, bar)`
- [ ] See how we can reduce boilerplate e.g. via a decorator in the view functions
